url3aa01d34ec_katsuster_ememu_tgz-pJ8-net_katsuster_ememu_ui_MainJ8/src/net/katsuster/ememu/arm/core/DecodeStageARMv5.java

```diff
@@ -0,0 +1,696 @@
+     * イミディエートを取るデータ処理命令に共通する命令をデコードします。
+     *
+     * 下記の種類の命令を扱います。
+     * and, eor, sub, rsb,
+     * add, adc, sbc, rsc,
+     * tst, teq, cmp, cmn,
+     * orr, mov, bic, mvn,
+     *
+     * @param inst ARM 命令
+     * @param id   オペコードフィールドと S ビットが示す演算の ID
+     * @return 命令の種類
+     */
+    public OpIndex decodeALUCommon(InstructionARM inst, int id) {
+        switch (id) {
+        case InstructionARM.OPCODE_S_AND:
+            return OpIndex.INS_ARM_ALUAND;
+        case InstructionARM.OPCODE_S_EOR:
+            return OpIndex.INS_ARM_ALUEOR;
+        case InstructionARM.OPCODE_S_SUB:
+            return OpIndex.INS_ARM_ALUSUB;
+        case InstructionARM.OPCODE_S_RSB:
+            return OpIndex.INS_ARM_ALURSB;
+        case InstructionARM.OPCODE_S_ADD:
+            return OpIndex.INS_ARM_ALUADD;
+        case InstructionARM.OPCODE_S_ADC:
+            return OpIndex.INS_ARM_ALUADC;
+        case InstructionARM.OPCODE_S_SBC:
+            return OpIndex.INS_ARM_ALUSBC;
+        case InstructionARM.OPCODE_S_RSC:
+            return OpIndex.INS_ARM_ALURSC;
+        case InstructionARM.OPCODE_S_TST:
+            return OpIndex.INS_ARM_ALUTST;
+        case InstructionARM.OPCODE_S_TEQ:
+            return OpIndex.INS_ARM_ALUTEQ;
+        case InstructionARM.OPCODE_S_CMP:
+            return OpIndex.INS_ARM_ALUCMP;
+        case InstructionARM.OPCODE_S_CMN:
+            return OpIndex.INS_ARM_ALUCMN;
+        case InstructionARM.OPCODE_S_ORR:
+            return OpIndex.INS_ARM_ALUORR;
+        case InstructionARM.OPCODE_S_MOV:
+            return OpIndex.INS_ARM_ALUMOV;
+        case InstructionARM.OPCODE_S_BIC:
+            return OpIndex.INS_ARM_ALUBIC;
+        case InstructionARM.OPCODE_S_MVN:
+            return OpIndex.INS_ARM_ALUMVN;
+        default:
+            throw new IllegalArgumentException("Unknown opcode S-bit ID " +
+                    String.format("%d.", id));
+        }
+    }
+
+    /**
+     * ロード、ストア命令をデコードします。
+     *
+     * subcode = 0b01
+     *
+     * 各ビットと命令の対応は下記の通りです。
+     *
+     *        |  I  |  P  |  B  |  W  |  L  |     |
+     *        | 25  | 24  | 22  | 21  | 20  |  4  |
+     * -------+-----+-----+-----+-----+-----+-----+
+     * LDR    |  x  |  x  |  0  |  x  |  1  |  x  |
+     * LDRB   |  x  |  x  |  1  |  x  |  1  |  x  |
+     * LDRBT  |  x  |  0  |  1  |  1  |  1  |  x  |
+     * LDRT   |  x  |  0  |  0  |  1  |  1  |  x  |
+     * -------+-----+-----+-----+-----+-----+-----+
+     * STR    |  x  |  x  |  0  |  x  |  0  |  x  |
+     * STRB   |  x  |  x  |  1  |  x  |  0  |  x  |
+     * STRBT  |  x  |  0  |  1  |  1  |  0  |  x  |
+     * STRT   |  x  |  0  |  0  |  1  |  0  |  x  |
+     * -------+-----+-----+-----+-----+-----+-----+
+     * UND    |  1  |  x  |  x  |  x  |  x  |  1  |
+     * -------+-----+-----+-----+-----+-----+-----+
+     *
+     * @param inst ARM 命令
+     * @return 命令の種類
+     */
+    public OpIndex decodeLdrStr(InstructionARM inst) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 21. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.